### PR TITLE
[12.0] Port base_exception fixes (locks, memory)

### PR DESCRIPTION
**Pull request by guewen**
Port of:

* https://github.com/OCA/server-tools/pull/1620
* https://github.com/OCA/server-tools/pull/1647
* https://github.com/OCA/server-tools/pull/1644

Description on https://github.com/OCA/server-tools/pull/1642

The commit messages and links above explain the issues.

### CODE REVIEW

**Review by florian-dacosta — APPROVED** [review]

File: base_exception/models/base_exception.py [hunk 1]
```diff
@@ -86,28 +86,44 @@ def _rule_domain(self):
...
+        # * write on write_date so lock the expection.rule
+        # * trigger the recomputation of "main_exception_id" on
+        #   all the sale orders related to the rule, locking them all
+        #   and preventing concurrent writes
+        # Reversing the write by writing on SaleOrder instead of
+        # ExceptionRule fixes the 2 kinds of unexpected locks.
+        # It should not result in more queries than writing on ExceptionRule:
+        # the "to remove" part generates one DELETE per rule on the relation table
+        # and the "to add" part generates one INSERT (with unnest) per rule.
+        for rule_id, records in rules_to_remove.iteritems():
```
Comment (florian-dacosta) [inline base_exception/models/base_exception.py]:
```suggestion
        for rule_id, records in rules_to_remove.items():
```

File: base_exception/models/base_exception.py [hunk 2]
```diff
@@ -86,28 +86,44 @@ def _rule_domain(self):
...
+            records.write({'exception_ids': [(3, rule_id,)]})
+        for rule_id, records in rules_to_add.iteritems():
```
Comment (florian-dacosta) [inline base_exception/models/base_exception.py]:
```suggestion
        for rule_id, records in rules_to_add.items():
```

**Review by florian-dacosta — CHANGES_REQUESTED** [review]
I actually approved too fast.
`iteritems` does not exist anymore in python 3

File: base_exception/models/base_exception.py [hunk 3]
```diff
@@ -86,28 +86,44 @@ def _rule_domain(self):
...
+        # before linking. We don't want to call "rule.write()"
+        # which would:
+        # * write on write_date so lock the expection.rule
+        # * trigger the recomputation of "main_exception_id" on
+        #   all the sale orders related to the rule, locking them all
+        #   and preventing concurrent writes
+        # Reversing the write by writing on SaleOrder instead of
+        # ExceptionRule fixes the 2 kinds of unexpected locks.
+        # It should not result in more queries than writing on ExceptionRule:
+        # the "to remove" part generates one DELETE per rule on the relation table
```
Comment (yvaucher) [inline base_exception/models/base_exception.py]: Warning there is a lint issue here

base_exception/models/base_exception.py:121:80: E501 line too long (82 > 79 characters)

Comment (guewen) [inline base_exception/models/base_exception.py]:
```suggestion
        # the "to remove" part generates one DELETE per rule on the relation
        # table
```

**Comment by guewen** [comment]
Squashed and rebased

**Review by sebastienbeau — APPROVED** [review]
LGTM

**Review by yvaucher — APPROVED** [review]

**Pull request merged**